### PR TITLE
MC-35 marqo GA changes

### PR DESCRIPTION
**Pull request by danyilq**
* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Support for marqo GA.

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**
when url is api.marqo.ai it uses resolver for operations related to indexes to get their url.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No.

* **Other information**:

### CODE REVIEW

**Comment by pandu-k** [comment]
hi @danyilq ! Have you ran `tox` unit tests? And do they pass?

File: src/marqo/client.py [hunk 1]
```diff
@@ -46,7 +46,9 @@ def create_index(
         sentences_per_chunk=2,
         sentence_overlap=0,
         image_preprocessing_method=None,
-        settings_dict=None
+        settings_dict=None,
+        inference_type="marqo.small",
+        storage_type="marqo.medium",
```
Comment (pandu-k) [inline src/marqo/client.py]: Are these defaults configured in the cloud somewhere? If so, I think we should leave these as `None` by default, and let the cloud determine defaults if not explicitly set by the user. 

Otherwise we will have defaults defined in 2 places.

Comment (pandu-k) [reply src/marqo/client.py]: Also, how about being even more transparent and explicit about what these parameters mean? How about
`inference_node_type` and `storage_node_type`

File: src/marqo/config.py [hunk 2]
```diff
@@ -43,5 +46,15 @@ def set_url(self, url):
             self.cluster_is_remote = True
             if "s2search.io" in lowered_url:
                 self.cluster_is_s2search = True
+            if "api.marqo.ai" in lowered_url:
+                self.cluster_is_marqo = True
+                self.marqo_url_resolver = MarqoUrlResolver(api_key=self.api_key, expiration_time=60)
         self.url = url
         return self.url
+
+    def get_url(self, index_name=None,):
```
Comment (pandu-k) [inline src/marqo/config.py:55]: Add a docstring about what this is for

File: src/marqo/index.py [hunk 3]
```diff
@@ -54,7 +56,9 @@ def create(config: Config, index_name: str,
                sentences_per_chunk=2,
                sentence_overlap=0,
                image_preprocessing_method=None,
-               settings_dict: dict = None
+               settings_dict: dict = None,
+               inference_type: str = "marqo.CPU",
```
Comment (pandu-k) [inline src/marqo/index.py]: why is both `"inference_type": "marqo.small",` and `inference_type: str = "marqo.CPU",` defined in the code in different places?

File: src/marqo/marqo_url_resolver.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import time
+
+import requests
+
+
+class MarqoUrlResolver:
```
Comment (pandu-k) [inline src/marqo/marqo_url_resolver.py:8]: Needs tests for each method, and tests around the changing object state.

File: src/marqo/index.py [hunk 5]
```diff
@@ -91,7 +97,17 @@ def create(config: Config, index_name: str,
...
+                return req.post(f"indexes/{index_name}", body=cl_settings)
+            cl_settings['inference_type'] = inference_type
+            cl_settings['storage_type'] = storage_type
+            req.post(f"indexes/{index_name}", body=cl_settings)
+            creation = req.get(f"indexes/{index_name}/status")
+            while creation['index_status'] != 'READY':
+                time.sleep(10)
+                creation = req.get(f"indexes/{index_name}/status")
+                mq_logger.info(f"Index creation status: {creation['status']}")
+            return creation
```
Comment (pandu-k) [inline src/marqo/index.py]: Needs tests for creating indexes for the cloud

File: src/marqo/config.py (hunk 2 above)
Comment (pandu-k) [reply src/marqo/config.py:55]: Simple tests about the behaviour would be good to prevent accidental future regressions

File: src/marqo/config.py [hunk 6]
```diff
@@ -43,5 +46,15 @@ def set_url(self, url):
...
+        if not self.cluster_is_marqo:
+            return self.url
+        if self.cluster_is_marqo and not index_name:
+            return self.url + "/api"
+        return self.marqo_url_resolver.urls_mapping[index_name]
```
Comment (pandu-k) [inline src/marqo/config.py]: do we need error handling if `index_name` not in `urls_mapping`

File: src/marqo/marqo_url_resolver.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    def __init__(self, api_key=None, expiration_time: int = 60):
+        """ URL Resolver is a cache for urls that are resolved to their respective indices only for marqo cloud. """
+        self.timestamp = time.time() - 60
```
Comment (pandu-k) [inline src/marqo/marqo_url_resolver.py]: consider `self.timestamp = time.time() - expiration_time ` if we consistently want to refresh the URLs after initialisation

File: src/marqo/config.py (hunk 2 above)
Comment (pandu-k) [reply src/marqo/config.py:55]: We should also note in the docstring that this method kicks off the URL refresh process

**Review by pandu-k — COMMENTED** [review]
One of the problems with this is that `HttpRequests. send_request()` (which is called on every HTTP request to Marqo - including latency critical search requests), periodically starts blocking calls to `refresh_urls_if_needed` before the  requested HTTP request be sent.

One way around this to make `refresh_urls_if_needed` execute in its own thread, unless the requested index is not found in `_urls_mapping`, in which case a blocking call is alright

File: src/marqo/index.py [hunk 8]
```diff
@@ -74,8 +75,9 @@ def create(config: Config, index_name: str,
             settings_dict: if specified, overwrites all other setting
                 parameters, and is passed directly as the index's
                 index_settings
-            inference_type: inference type for the index
-            storage_type: storage type for the index
+            inference_node_type: inference type for the index
+            storage_node_type: storage type for the index
+            inference_node_count: number of inference nodes for the index
```
Comment (pandu-k) [inline src/marqo/index.py:80]: Can you please create unit tests for `inference_node_type`, `inference_node_count` and `storage_node_type` parameters, including edge cases (`None` or `0`)

File: tests/v0_tests/test_index.py [hunk 9]
```diff
@@ -200,4 +200,13 @@ def test_create_custom_number_of_replicas(self):
         self.client.create_index(index_name=self.index_name_1, settings_dict = settings)
         index_setting = self.client.index(self.index_name_1).get_settings()
         print(index_setting)
-        assert intended_replicas == index_setting['number_of_replicas']
\ No newline at end of file
+        assert intended_replicas == index_setting['number_of_replicas']
+
+    @mock.patch("marqo._httprequests.HttpRequests.post", return_value="Record added to DynamoDB table successfully")
+    @mock.patch("marqo._httprequests.HttpRequests.get", return_value={"index_status": "READY"})
+    def test_create_marqo_cloud_index(self, mock_post, mock_get):
```
Comment (pandu-k) [inline tests/v0_tests/test_index.py]: check that URL string params are as expected

File: src/marqo/index.py [hunk 10]
```diff
@@ -54,7 +56,10 @@ def create(config: Config, index_name: str,
                sentences_per_chunk=2,
                sentence_overlap=0,
                image_preprocessing_method=None,
-               settings_dict: dict = None
+               settings_dict: dict = None,
+               inference_node_type: str = None,
+               storage_node_type: str = None,
+               inference_node_count: int = 1,
```
Comment (pandu-k) [inline src/marqo/index.py:62]: should also let the API decide what the default `inference_node_count` is?

Comment (danyilq) [reply src/marqo/index.py:62]: this value must be greater than 0 anyways, api will validate the value, but I think that having it set to 1 in the client is reasonable so that customer doesn't get unexpected value by api. Whilst forcing customer to choose specific storage and inference types.

File: src/marqo/config.py (hunk 6 above)
Comment (pandu-k) [reply src/marqo/config.py]: There's no guarantee that the index_name is going to end up in mappings right? So for a non existent index, they won't get a `404 index not found` from Marqo,  they will get a KeyError

File: src/marqo/marqo_url_resolver.py [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+class MarqoUrlResolver:
+    def __init__(self, api_key=None, expiration_time: int = 60):
+        """ URL Resolver is a cache for urls that are resolved to their respective indices only for marqo cloud. """
+        self.timestamp = time.time() - expiration_time
+        self._urls_mapping = {}
+        self.api_key = api_key
+        self.expiration_time = expiration_time
+
+    def refresh_urls_if_needed(self, index_name):
+        if index_name not in self._urls_mapping and time.time() - self.timestamp > self.expiration_time:
```
Comment (pandu-k) [inline src/marqo/marqo_url_resolver.py]: This basically means that we are OK with an index_name's url _never_ being updated in this client once set (including if it has been deleted and recreated). Are we sure we are alright with this?

File: src/marqo/marqo_url_resolver.py [hunk 12]
```diff
@@ -0,0 +1,30 @@
+import time
+
+import requests
+
+
+class MarqoUrlResolver:
+    def __init__(self, api_key=None, expiration_time: int = 60):
```
Comment (pandu-k) [inline src/marqo/marqo_url_resolver.py]: nit: Are we sure that we are willing to only check for updates after a minute? This feels a little long. For example, someone else creates an index, I will have to wait up to a minute after it is available before I can use it. Perhaps a lower value, like 15 seconds or so?

Comment (danyilq) [reply src/marqo/marqo_url_resolver.py]: It sounds reasonable, I agree

File: src/marqo/marqo_url_resolver.py (hunk 11 above)
Comment (danyilq) [reply src/marqo/marqo_url_resolver.py]: how about we refresh it once in 5-15 minutes if index is already presented, if customer were to delete and recreate their deletion would take around 5 minutes and after <=(5-15) minutes their index will be deleted from urls list and then each time they request it would refresh at the same rate as we have default.

File: src/marqo/config.py (hunk 6 above)
Comment (danyilq) [reply src/marqo/config.py]: Mindset was to keep client dumb and let api handle it, but in that case we indeed need to handle it with client, wip on that one

Comment (danyilq) [reply src/marqo/config.py]: Is it reasonable to raise Exception(f"Index {item} is still being created") when index is creating
and Exception(f"Index {item} does not exist") or should it be softer and not stopping the code?

Comment (pandu-k) [reply src/marqo/config.py:64]: Yes I agree that we need to keep the client as simple as possible, but we may have no choice here. 

In terms of error message we should be as clear as possible.

For example something like this (feel free to modify this if needed) :

"The Python client could not resolve the endpoint for the index name `my-index`. This could be due to the index is still in the process of being created, or that the client's cache has not yet been updated. 
- please try again in a couple of minutes, to allow the client's cache to be updated
- you can also try the querying the index's status: GET https://<INSERT INDEX ENDPOINT HERE>/status
- If the issue still persists please make a Marqo cloud support ticket here <LINK

Comment (pandu-k) [reply src/marqo/config.py]: Yes I agree that we need to keep the client as simple as possible, but we may have no choice here. 

In terms of error message we should be as clear as possible.

For example something like this (feel free to modify this if needed) :
``` 
The Python client could not resolve the endpoint for the index name `my-index`. 
This could be due to the index is still in the process of being created, or that the client's cache has not yet been updated. 
- please try again in a couple of minutes, to allow the client's cache to be updated
- you can also try the querying the index's status: `curl https://<INSERT INDEX ENDPOINT HERE>/status`
- If the issue still persists please make a Marqo cloud support ticket here <LINK>
```

File: src/marqo/errors.py [hunk 13]
```diff
@@ -165,3 +165,27 @@ class BackendTimeoutError(InternalError):
...
+    """Error when Marqo index is not ready"""
+    code = "index_not_ready_cloud"
+    status_code = HTTPStatus.NOT_FOUND
+
+    def __init__(self, index_name: str,) -> None:
+        self.message = f"The Python client could not resolve the endpoint for the index name {index_name}. " \
+                       f"This could be due to the index is still in the process of being created," \
+                       f" or that the client's cache has not yet been updated.\n" \
+                       f"- Please try again in a couple of minutes, " \
+                       f"or you can query the index status with get_status function to see when index is ready.\n" \
```
Comment (pandu-k) [inline src/marqo/errors.py]: `mq.('my-index').get_status()`

Comment (pandu-k) [reply src/marqo/errors.py]: - Consider subclassing this, because it's a client specific, rather than HTTP errror: https://github.com/marqo-ai/py-marqo/blob/ab2c3de94f02ad0ef8440da743f6e18600f701a1/src/marqo/errors.py#L6

File: src/marqo/errors.py [hunk 14]
```diff
@@ -165,3 +165,27 @@ class BackendTimeoutError(InternalError):
     def __init__(self, message: str,) -> None:
         self.message = f"Timeout error communicating with Marqo: {message}"
 
+
+class MarqoIndexNotReadyError(__InvalidRequestError):
```
Comment (pandu-k) [inline src/marqo/errors.py]: Add `Cloud` to error name: `MarqoIndexNotReadyErrorCloud`

File: src/marqo/errors.py [hunk 15]
```diff
@@ -165,3 +165,27 @@ class BackendTimeoutError(InternalError):
...
+    def __init__(self, index_name: str,) -> None:
+        self.message = f"The Python client could not resolve the endpoint for the index name {index_name}. " \
+                       f"This could be due to the index is still in the process of being created," \
+                       f" or that the client's cache has not yet been updated.\n" \
+                       f"- Please try again in a couple of minutes, " \
+                       f"or you can query the index status with get_status function to see when index is ready.\n" \
+                       f"- If the problem persists, please contact marqo support at support@marqo.ai"
+
+
+class MarqoIndexNotFoundError(__InvalidRequestError):
```
Comment (pandu-k) [inline src/marqo/errors.py]: ditto RE "Add Cloud to error name"

File: src/marqo/marqo_url_resolver.py [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+        self.refresh_urls_if_needed(item)
+        if item in self._urls_mapping['READY']:
+            return self._urls_mapping['READY'][item]
+        if item in self._urls_mapping['CREATING']:
+            raise MarqoIndexNotReadyError(item)
+        raise MarqoIndexNotFoundError(item)
+
+    def _refresh_urls(self):
+        response = requests.get('https://api.marqo.ai/api/indexes',
+                                headers={"x-api-key": self.api_key}).json()
```
Comment (pandu-k) [inline src/marqo/marqo_url_resolver.py]: Consider logging a warning if response is not OK

File: src/marqo/marqo_url_resolver.py [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+        self.api_key = api_key
+        self.expiration_time = expiration_time
+
+    def refresh_urls_if_needed(self, index_name):
+        if index_name not in self._urls_mapping['READY'] and time.time() - self.timestamp > self.expiration_time:
+            # fast refresh to catch if index was created
+            self._refresh_urls()
+        if index_name in self._urls_mapping['READY'] and time.time() - self.timestamp > 360:
+            # slow refresh in case index was deleted
+            self._refresh_urls()
```
Comment (pandu-k) [inline src/marqo/marqo_url_resolver.py]: Consider adding a timeout here, for the slow case (we don't want to hold up search/add docs requests if the control plane is acting badly). Perhaps 3 seconds

**Review by pandu-k — COMMENTED** [review]
Nice!
+ make a few tests around index_name within HTTP requests (to prevent future regressions)
  + this is because it isn't intuitive that `http.get(url, index_name=index_name)` triggers Marqo-cloud specific functionality, when `index_name` is set
  + You may want to add a docstring comment to explain this non-intuitive functionality 
+ tests that it doesn't use the API resolver if the URL isn't api.marqo

**Review by pandu-k — APPROVED** [review]
